projects/angular/src/forms/datepicker/date-single-input.ts

```diff
@@ -0,0 +1,145 @@
+  onValueChange(target: HTMLInputElement) {
+    const validDateValue = this.dateIOService.getDateValueFromDateString(target.value);
+    if (this.usingClarityDatepicker() && validDateValue) {
+      this.updateDate(validDateValue, true);
+    } else if (this.usingNativeDatepicker()) {
+      const [year, month, day] = target.value.split('-');
+      this.updateDate(new Date(+year, +month - 1, +day), true);
+    } else {
+      this.emitDateOutput(null);
+    }
+  }
+
+  override ngOnInit() {
+    super.ngOnInit();
+
+    this.subscriptions.push(this.listenForUserSelectedDayChanges(), this.listenForControlValueChanges());
+  }
+
+  ngAfterViewInit() {
+    // I don't know why I have to do this but after using the new HostWrapping Module I have to delay the processing
+    // of the initial Input set by the user to here. If I do not 2 issues occur:
+    // 1. The Input setter is called before ngOnInit. ngOnInit initializes the services without which the setter fails.
+    // 2. The Renderer doesn't work before ngAfterViewInit (It used to before the new HostWrapping Module for some reason).
+    // I need the renderer to set the value property on the input to make sure that if the user has supplied a Date
+    // input object, we reflect it with the right date on the input field using the IO service. I am not sure if
+    // these are major issues or not but just noting them down here.
+    this.processInitialInputs();
+  }
+
+  protected updateDate(value: Date, setByUserInteraction = false) {
+    const date = this.getValidDateValueFromDate(value);
+    if (setByUserInteraction) {
+      this.emitDateOutput(date);
+    } else {
+      this.previousDateChange = date;
+    }
+
+    if (this.dateNavigationService) {
+      this.dateNavigationService.persistedDate = this.dateNavigationService.selectedDay = date
+        ? new DayModel(date.getFullYear(), date.getMonth(), date.getDate())
+        : null;
+    }
+
+    this.updateInput(date);
+  }
+
+  private listenForControlValueChanges() {
+    return of(this.datepickerHasFormControl())
+      .pipe(
+        filter(hasControl => hasControl),
+        switchMap(() => this.control.valueChanges),
+        // only update date value if not being set by user
+        filter(() => !this.datepickerFocusService.elementIsFocused(this.el.nativeElement))
+      )
+      .subscribe((value: string) => this.updateDate(this.dateIOService.getDateValueFromDateString(value)));
+  }
+
+  private listenForUserSelectedDayChanges() {
+    return this.dateNavigationService.selectedDayChange.subscribe(dayModel => {
+      this.updateDate(dayModel.toDate(), true);
+    });
+  }
+
+  private emitDateOutput(date: Date) {
+    if (!datesAreEqual(date, this.previousDateChange)) {
+      this.dateChange.emit(date);
+      this.previousDateChange = date;
+    } else if (!date && this.previousDateChange) {
+      this.dateChange.emit(null);
+      this.previousDateChange = null;
+    }
+  }
```

Comment:
This method is repeated in all three base classes. You can add a `protected abstract dateChange` property to the base class for the event and move the `previousDateChange` property to the base class. See 4919ceaf769f59043954d6054ae6f2500931093a for and example of how this can be done.

Reply:
I will provide an example commit that moves all of the duplicated code back to the base class.